lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp

```diff
@@ -3206,6 +3256,12 @@ static bool ReplaceMemcpy(Value *V, Value *Src, MemCpyInst *MC,
   if (Instruction *SrcI = dyn_cast<Instruction>(Src))
     if (!DominateAllUsers(SrcI, V, DT))
       return false;
+
+  // Unless we encounter an issue, we always replace V by Src below.
+  // Conservatively remove all lifetime users of V. Otherwise, Src would inherit
+  // them from V and suddenly have a different lifetime than it had before.
+  removeLifetimeUsers(V);
```

Comment:
If V is read outside of Src's lifetime (V's lifetime extends beyond Src's lifetime), the lifetime of Src needs to be expanded to include V.

Reply:
Hmm, the intention of this was to be conservative. I think I experimented with expanding lifetimes but hit some blockers.

If the code below actually performs the replacement of Src by V then removing lifetime intrinsics of V should be fine. The intrinsics of Src are dead code afterwards (they still exist, but there are no uses of the memory that they refer to anymore). So the result is that neither memory described by Src nor V has lifetime intrinsics attached afterwards, which is conservative.

Now that leaves the case where the code below ends up NOT doing the replacement. IIUC there's only two cases where this happens:
1. if V is not a constant but V == Src
2. if V is not a constant and the memcpy is not an unbounded array memcpy and Src and V are pointers to structs of identical layouts and either Src or V has more than one use.

Again, removing lifetime intrinsics of V should be conservative here. Since both V and Src are still separate this should not have any effect on Src. If lifetimes of Src are somehow wrong then the problem was there before already.

I must be missing something...

Reply:
I see what the problem is.

V is actually the one being replaced by Src, not the other way around. `removeLifetimeUsers` removed all of V's lifetime markers, but when V is removed and all its usage is replaced by Src. Now Src's lifetime markers stick around and are wrong.

But even if we do `removeLifetimeUsers` on Src instead, by doing `replaceAllUsesWith`, V's lifetime markers are now pointing to Src, which is still wrong. To achieve the intended result, we must remove the lifetime markers for both pointers.

But removing it here is maybe overly conservative, not all the cases below replace the pointers. Here's a patch:
```
diff --git a/lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp b/lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp
index 33ca05ee3..70a081ad6 100644
--- a/lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp
+++ b/lib/Transforms/Scalar/ScalarReplAggregatesHLSL.cpp
@@ -3237,6 +3237,13 @@ static void removeLifetimeUsers(Value *V) {
   }
 }
 
+// Should be merging the lifetime, but for convenience, just
+// remove them now to be safe.
+static void updateLifetimeForReplacement(Value *From, Value *To) {
+  removeLifetimeUsers(From);
+  removeLifetimeUsers(To);
+}
+
 static bool DominateAllUsers(Instruction *I, Value *V, DominatorTree *DT);
 
 
@@ -3257,14 +3264,10 @@ static bool ReplaceMemcpy(Value *V, Value *Src, MemCpyInst *MC,
     if (!DominateAllUsers(SrcI, V, DT))
       return false;
 
-  // Unless we encounter an issue, we always replace V by Src below.
-  // Conservatively remove all lifetime users of V. Otherwise, Src would inherit
-  // them from V and suddenly have a different lifetime than it had before.
-  removeLifetimeUsers(V);
-
   Type *TyV = V->getType()->getPointerElementType();
   Type *TySrc = Src->getType()->getPointerElementType();
   if (Constant *C = dyn_cast<Constant>(V)) {
+    updateLifetimeForReplacement(V, Src);
     if (TyV == TySrc) {
       if (isa<Constant>(Src)) {
         V->replaceAllUsesWith(Src);
@@ -3280,8 +3283,10 @@ static bool ReplaceMemcpy(Value *V, Value *Src, MemCpyInst *MC,
     }
   } else {
     if (TyV == TySrc) {
-      if (V != Src)
+      if (V != Src) {
+        updateLifetimeForReplacement(V, Src);
         V->replaceAllUsesWith(Src);
+      }
     } else if (!IsUnboundedArrayMemcpy(TyV, TySrc)) {
       Value* DestVal = MC->getRawDest();
       Value* SrcVal = MC->getRawSource();
@@ -3322,6 +3327,7 @@ static bool ReplaceMemcpy(Value *V, Value *Src, MemCpyInst *MC,
             MemcpySplitter::SplitMemCpy(MC, DL, annotation, typeSys);
             return true;
           } else {
+            updateLifetimeForReplacement(V, Src);
             DstPtr->replaceAllUsesWith(SrcPtr);
           }
         } else {
@@ -3330,6 +3336,7 @@ static bool ReplaceMemcpy(Value *V, Value *Src, MemCpyInst *MC,
         }
       }
     } else {
+      updateLifetimeForReplacement(V, Src);
       DXASSERT(IsUnboundedArrayMemcpy(TyV, TySrc), "otherwise mismatched types in memcpy are not unbounded array");
       ReplaceUnboundedArrayUses(V, Src);
     }
```

Reply:
Oh, I was getting confused by what is being replaced by what, sorry for that.
Yes, you are right, contrary to my comment in the code, removing only the lifetimes of V is not conservative. Replacing the memcpy effectively merges memory accesses of two previously disjoint allocas into a single alloca. If the allocas previously had non-equal lifetimes, leaving the lifetimes of Src intact can be wrong just the same way as leaving those of V intact (which apparently happens more frequently since I did encounter this case).
I also agree with your patch. It's a bit ugly to have multiple places where we remove lifetimes, but it's clearly better than unconditionally removing them, and I prefer this over restructuring the rest of that code to pull the cases that don't actually replace anything out.
I'll apply the patch and will also add a test for this.